File: src/main/java/com/violeth/blockjs/blockjs/jsinterface/mcinterface/World.java
package com.violeth.blockjs.blockjs.jsinterface.mcinterface;

import com.eclipsesource.v8.V8Function;
import com.violeth.blockjs.blockjs.BlockJS;

public class World {
    public int addBlockDamageListener(int x, int y, int z, V8Function callback) {
        return BlockJS.instance.binds.addBlockDamageEventListener(x, y, z, callback.twin());
    }
    public void removeBlockDamageListener(int id) {
        BlockJS.instance.binds.removeBlockDamageEventListener(id);
    }

    public int addBlockBreakListener(int x, int y, int z, V8Function callback) {
        return BlockJS.instance.binds.addBlockBreakEventListener(x, y, z, callback.twin());
    }
    public void removeBlockBreakListener(int id) {
        BlockJS.instance.binds.removeBlockBreakEventListener(id);
    }

    public int addBlockInteractListener(int x, int y, int z, V8Function callback) {
        return BlockJS.instance.binds.addBlockInteractEventListener(x, y, z, callback.twin());
    }
    public void removeBlockInteractListener(int id) {
        BlockJS.instance.binds.removeBlockInteractEventListener(id);
    }
}

File: plugin/src/main/kotlin/io/github/blockjs/plugin/js/runtime/api/game/World.kt
package io.github.blockjs.plugin.js.runtime.api.game

import com.caoccao.javet.annotations.V8Function
import com.caoccao.javet.exceptions.JavetException
import com.caoccao.javet.values.reference.V8ValueFunction
import io.github.blockjs.plugin.Main
import org.bukkit.Bukkit

@Suppress("unused")
object World {
    @V8Function
    fun getPlayerCount(): Int {
        return Bukkit.getOnlinePlayers().size
    }

    @V8Function
    fun addBlockDamageListener(x: Int, y: Int, z: Int, callback: V8ValueFunction): Int {
        try {
            callback.setWeak()

            return Main.instance!!.binds.addBlockDamageEventListener(x, y, z, callback)
        } catch (e: JavetException) {
            throw RuntimeException(e)
        }
    }

    @V8Function
    fun removeBlockDamageListener(id: Int) {
        Main.instance!!.binds.removeBlockDamageEventListener(id)
    }

    @V8Function
    fun addBlockBreakListener(x: Int, y: Int, z: Int, callback: V8ValueFunction): Int {
        try {
            callback.setWeak()

            return Main.instance!!.binds.addBlockBreakEventListener(x, y, z, callback.toClone())
        } catch (e: JavetException) {
            throw RuntimeException(e)
        }
    }

    @V8Function
    fun removeBlockBreakListener(id: Int) {
        Main.instance!!.binds.removeBlockBreakEventListener(id)
    }
}
